src/map/battle.cpp

```diff
@@ -977,62 +977,218 @@ static void battle_absorb_damage(struct block_list *bl, struct Damage *d) {
+		struct skill_unit_group *group = skill_id2group(sce->val3);
 
 		if (group) {
-			if (skill_id_val == MH_STEINWAND) {
-				if (--group->val2 <= 0)
+			d->dmg_lv = ATK_BLOCK;
+
+			switch (sce->val2) {
+			case MG_SAFETYWALL:
+				if (--group->val2 <= 0) {
 					skill_delunitgroup(group);
-				d->dmg_lv = ATK_BLOCK;
-				if( (group->val3 - damage) > 0 )
+					break;
+				}
+#ifdef RENEWAL
+				if ((group->val3 - damage) > 0)
 					group->val3 -= (int)cap_value(damage, INT_MIN, INT_MAX);
 				else
 					skill_delunitgroup(group);
-				return false;
+#endif
+				break;
+			case MH_STEINWAND:
+				if (--group->val2 <= 0) {
+					skill_delunitgroup(group);
+					break;
+				}
+				if ((group->val3 - damage) > 0)
+					group->val3 -= (int)cap_value(damage, INT_MIN, INT_MAX);
+				else
+					skill_delunitgroup(group);
+				break;
 			}
-			//in RE, SW possesses a lifetime equal to group val2, (3x caster hp, or homon formula)
+			return false;
+		}
+		status_change_end(target, SC_SAFETYWALL, INVALID_TIMER);
+	}
+
+	if ((sc->data[SC_PNEUMA] && (flag&(BF_MAGIC | BF_LONG)) == BF_LONG) ||
+		(sc->data[SC_BASILICA] && !status_bl_has_mode(src, MD_STATUS_IMMUNE)) ||
+		(sc->data[SC_ZEPHYR] && !(flag&BF_MAGIC && skill_id) && !(skill_get_inf(skill_id)&(INF_GROUND_SKILL | INF_SELF_SKILL))) ||
+		sc->data[SC__MANHOLE] ||
+		sc->data[SC_KINGS_GRACE]
+		)
+	{
+		d->dmg_lv = ATK_BLOCK;
+		return false;
+	}
+
+	if (sc->data[SC_WHITEIMPRISON]) { // Gravitation and Pressure do damage without removing the effect
+		if (skill_id == MG_NAPALMBEAT ||
+			skill_id == MG_SOULSTRIKE ||
+			skill_id == WL_SOULEXPANSION ||
+			(skill_id && skill_get_ele(skill_id, skill_lv) == ELE_GHOST) ||
+			(!skill_id && (status_get_status_data(src))->rhw.ele == ELE_GHOST))
+		{
+			if (skill_id == WL_SOULEXPANSION)
+				damage <<= 1; // If used against a player in White Imprison, the skill deals double damage.
+			status_change_end(target, SC_WHITEIMPRISON, INVALID_TIMER); // Those skills do damage and removes effect
+		} else {
 			d->dmg_lv = ATK_BLOCK;
-#ifdef RENEWAL
-			if ( ( group->val2 - damage) > 0 ) {
-				group->val2 -= (int)cap_value(damage, INT_MIN, INT_MAX);
-			} else
-				skill_delunitgroup(group);
 			return false;
-#else
-			if (--group->val2 <= 0)
-				skill_delunitgroup(group);
+		}
+	}
+
+	if ((sce = sc->data[SC_WEAPONBLOCKING]) && flag&(BF_SHORT | BF_WEAPON) && rnd() % 100 < sce->val2) {
+		clif_skill_nodamage(target, src, GC_WEAPONBLOCKING, sce->val1, 1);
+		sc_start2(src, target, SC_COMBO, 100, GC_WEAPONBLOCKING, src->id, skill_get_time2(GC_WEAPONBLOCKING, sce->val1));
+		d->dmg_lv = ATK_BLOCK;
+		return false;
+	}
+
+	if ((sce = sc->data[SC_MILLENNIUMSHIELD]) && sce->val2 > 0 && damage > 0) {
+		sce->val3 -= (int)cap_value(damage, INT_MIN, INT_MAX); // absorb damage
+		d->dmg_lv = ATK_BLOCK;
+		if (sce->val3 <= 0) { // Shield Down
+			sce->val2--;
+			if (sce->val2 >= 0) {
+				clif_millenniumshield(target, sce->val2);
+				if (!sce->val2)
```

Comment:
I think this a very bad way of expressing your intention, since  `val2` is a integer. You should just do `val2 != 0` or whatever you are trying to do. Write it more explicitly.